### PR TITLE
62 accept and output standardized stdin and stdout

### PR DESCRIPTION
**Pull request by garrettthomaskth**

### CODE REVIEW

File: esque/cli/commands.py [hunk 1]
```diff
@@ -127,8 +152,14 @@ def create_topic(state: State, topic_name: str, like=None):
 @click.argument("topic-name", required=True)
 @pass_state
 def edit_topic(state: State, topic_name: str):
+
+    if not sys.__stdin__.isatty():
```
Comment (swenzel) [inline esque/cli/commands.py]:
```suggestion
    if not click.get_text_stream("stdin").isatty():
```

File: esque/cli/output.py [hunk 2]
```diff
@@ -213,3 +218,32 @@ def green_bold(s: str) -> str:
...
+
+class BytesEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, bytes):
+            return obj.decode("UTF-8")
+            # Let the base class default method raise the TypeError
+        return json.JSONEncoder.default(self, obj)
+
+
+def format_output(output, output_format):
```
Comment (swenzel) [inline esque/cli/output.py]:
```suggestion
def format_output(output: Any, output_format: str) -> str:
```

File: esque/resources/broker.py [hunk 3]
```diff
@@ -32,7 +32,7 @@ def get_all(cls, cluster) -> List["Broker"]:
         return sorted(brokers, key=operator.attrgetter("broker_id"))
 
     @translate_third_party_exceptions
-    def describe(self):
+    def describe(self) -> dict:
```
Comment (swenzel) [inline esque/resources/broker.py]:
```suggestion
    def describe(self) -> Dict:
```

File: esque/cli/commands.py [hunk 4]
```diff
@@ -222,19 +246,22 @@ def apply(state: State, file: str):
...
 @describe.command("topic")
-@click.argument("topic-name", required=True, type=click.STRING, autocompletion=list_topics)
+@click.argument(
+    "topic-name", callback=fallback_to_stdin, required=False, type=click.STRING, autocompletion=list_topics
+)
+@output_format_option
 @error_handler
 @pass_state
-def describe_topic(state, topic_name):
+def describe_topic(state, topic_name, output_format):
```
Comment (swenzel) [inline esque/cli/commands.py]:
```suggestion
def describe_topic(state: State, topic_name: str, output_format: str):
```

File: esque/cli/commands.py [hunk 5]
```diff
@@ -251,16 +278,17 @@ def get_offsets(state, topic_name):
 
 
 @describe.command("broker")
-@click.argument("broker-id", required=True)
+@click.argument("broker-id", callback=fallback_to_stdin, required=False)
 @error_handler
 @pass_state
-def describe_broker(state, broker_id):
+@output_format_option
+def describe_broker(state, broker_id, output_format):
```
Comment (swenzel) [inline esque/cli/commands.py]:
```suggestion
def describe_broker(state: State, broker_id: str, output_format: str):
```

File: esque/cli/commands.py [hunk 6]
```diff
@@ -269,11 +297,12 @@ def describe_broker(state, broker_id):
 )
 @error_handler
 @pass_state
-def describe_consumergroup(state, consumer_id, all_partitions):
+@output_format_option
+def describe_consumergroup(state, consumer_id, all_partitions, output_format):
```
Comment (swenzel) [inline esque/cli/commands.py]:
```suggestion
def describe_consumergroup(state: State, consumer_id: str, all_partitions: bool, output_format: str):
```

File: esque/cli/commands.py [hunk 7]
```diff
@@ -295,11 +324,11 @@ def get_consumergroups(state):
 
 
 @get.command("topics")
-@click.argument("topic", required=False, type=click.STRING, autocompletion=list_topics)
+@click.option("-p", "--prefix", type=click.STRING, autocompletion=list_topics, required=False)
 @error_handler
 @pass_state
-def get_topics(state, topic):
-    topics = state.cluster.topic_controller.list_topics(search_string=topic, get_topic_objects=False)
+def get_topics(state, prefix):
```
Comment (swenzel) [inline esque/cli/commands.py]:
```suggestion
def get_topics(state: State, prefix: str):
```

File: tests/conftest.py [hunk 8]
```diff
@@ -41,7 +44,14 @@ def pytest_collection_modifyitems(config, items):
 
 
 @pytest.fixture()
-def cli_runner():
+def interactive_cli_runner(test_config, monkeypatch):
```
Comment (swenzel) [inline tests/conftest.py]:
```suggestion
def interactive_cli_runner(test_config: Config, monkeypatch: MonkeyPatch):
```

File: tests/conftest.py [hunk 9]
```diff
@@ -41,7 +44,14 @@ def pytest_collection_modifyitems(config, items):
...
+    monkeypatch.setattr(click_stdin, "isatty", lambda: True)
+    yield CliRunner()
+
+
+@pytest.fixture()
+def non_interactive_cli_runner(test_config, monkeypatch):
```
Comment (swenzel) [inline tests/conftest.py]:
```suggestion
def non_interactive_cli_runner(test_config: Config, monkeypatch: MonkeyPatch):
```

File: tests/conftest.py [hunk 10]
```diff
@@ -217,3 +233,9 @@ def cluster(test_config):
 @pytest.fixture()
 def state(test_config):
     yield State()
+
+
+def check_and_load_yaml(output: str) -> dict:
```
Comment (swenzel) [inline tests/conftest.py]:
```suggestion
def check_and_load_yaml(output: str) -> Dict:
```

File: tests/integration/commands/test_creation.py [hunk 11]
```diff
@@ -8,17 +8,81 @@
...
+
+    topics = confluent_admin_client.list_topics(timeout=5).topics.keys()
+    assert topic_id not in topics
+
+
+@pytest.mark.integration
+def test_create_topic_without_topic_name_fails(non_interactive_cli_runner: CliRunner):
+    result = non_interactive_cli_runner.invoke(create_topic)
+    assert result.exit_code == 1
+    assert "ERROR: Missing argument TOPIC_NAME" in result.output
```
Comment (swenzel) [inline tests/integration/commands/test_creation.py]: Better to use `--verbose` and check for Exception type. If you check for the exact string, the test is too specific.
There are a few more places that I commented with 'exception type'.

File: tests/integration/commands/test_creation.py [hunk 12]
```diff
@@ -8,17 +8,81 @@
...
+
+@pytest.mark.integration
+def test_topic_creation_stops_in_non_interactive_mode_without_no_verify(
+    non_interactive_cli_runner: CliRunner, confluent_admin_client: confluent_kafka.admin.AdminClient, topic_id: str
+):
+    topics = confluent_admin_client.list_topics(timeout=5).topics.keys()
+    assert topic_id not in topics
+
+    result = non_interactive_cli_runner.invoke(create_topic, input=topic_id)
+    assert (
```
Comment (swenzel) [inline tests/integration/commands/test_creation.py]: exception type

File: tests/integration/commands/test_deletion.py [hunk 13]
```diff
@@ -16,13 +16,36 @@ def duplicate_topic(num_partitions, topic_factory):
...
+    confluent_admin_client.poll(timeout=1)
+    topics = confluent_admin_client.list_topics(timeout=5).topics.keys()
+    assert topic in topics
+
+
+@pytest.mark.integration
+def test_delete_topic_without_topic_name_fails(interactive_cli_runner: CliRunner):
+    result = interactive_cli_runner.invoke(delete_topic)
+    assert result.exit_code == 1
+    assert "ERROR: Missing argument TOPIC_NAME" in result.output
```
Comment (swenzel) [inline tests/integration/commands/test_deletion.py]: exception type

File: tests/integration/commands/test_deletion.py [hunk 14]
```diff
@@ -31,9 +54,45 @@ def test_topic_deletion_works(
...
+
+@pytest.mark.integration
+def test_topic_deletion_stops_in_non_interactive_mode_without_no_verify(
+    non_interactive_cli_runner: CliRunner, confluent_admin_client: confluent_kafka.admin.AdminClient, topic: str
+):
+    topics = confluent_admin_client.list_topics(timeout=5).topics.keys()
+    assert topic in topics
+
+    result = non_interactive_cli_runner.invoke(delete_topic, input=topic)
+    assert (
```
Comment (swenzel) [inline tests/integration/commands/test_deletion.py]: exception type

File: tests/integration/commands/test_describe.py [hunk 15]
```diff
@@ -1,11 +1,118 @@
...
+    assert result.exit_code == 0
+    output_dict = loader(result.output)
+    check_described_topic(output_dict)
+
+
+@pytest.mark.integration
+def test_describe_topic_without_topic_name_fails(non_interactive_cli_runner: CliRunner):
+    result = non_interactive_cli_runner.invoke(describe_topic)
+    assert result.exit_code == 1
+    assert "ERROR: Missing argument TOPIC_NAME" in result.output
```
Comment (swenzel) [inline tests/integration/commands/test_describe.py]: exception type

File: tests/integration/commands/test_describe.py [hunk 16]
```diff
@@ -1,11 +1,118 @@
...
+    check_described_broker(output_dict)
 
 
 @pytest.mark.integration
-def test_smoke_test_describe_topic(cli_runner: CliRunner, topic: str):
-    result = cli_runner.invoke(describe_topic, [topic])
+def test_describe_broker_without_broker_id_fails(non_interactive_cli_runner: CliRunner):
+    result = non_interactive_cli_runner.invoke(describe_broker)
+    assert result.exit_code == 1
+    assert "ERROR: Missing argument BROKER_ID" in result.output
```
Comment (swenzel) [inline tests/integration/commands/test_describe.py]: exception type

File: tests/integration/commands/test_edit.py [hunk 17]
```diff
@@ -49,12 +51,19 @@ def test_topic_creation_with_template_works(
...
 
     topic_config_dict = topic_controller.get_cluster_topic(topic).as_dict(only_editable=True)
     assert topic_config_dict == config_dict
+
+
+@pytest.mark.integration
+def test_edit_topic_without_topic_name_fails(non_interactive_cli_runner: CliRunner):
+    result = non_interactive_cli_runner.invoke(edit_topic)
+    assert result.exit_code != 0
+    assert 'Error: Missing argument "TOPIC_NAME"' in result.output
```
Comment (swenzel) [inline tests/integration/commands/test_edit.py]: exception type

File: tests/integration/commands/test_get.py [hunk 18]
```diff
@@ -1,11 +1,55 @@
...
-from esque.cli.commands import get_topics
+from esque.cli.commands import get_topics, get_consumergroups, get_brokers
+from esque.controller.topic_controller import TopicController
+from esque.resources.topic import Topic
+
+
+@pytest.mark.integration
+def test_smoke_test_get_topics(non_interactive_cli_runner: CliRunner):
+    result = non_interactive_cli_runner.invoke(get_topics, ["--verbose"])
+    assert "No config provided in" not in result.output
```
Comment (swenzel) [inline tests/integration/commands/test_get.py]: checking for `result.exit_code == 0` should be sufficient. If anything, you should make sure that no exception whatsoever was thrown by `assert result.exception is None`.

File: tests/integration/commands/test_get.py [hunk 19]
```diff
@@ -1,11 +1,55 @@
...
+    non_interactive_cli_runner: CliRunner,
+    topic_controller: TopicController,
+    confluent_admin_client: confluent_kafka.admin.AdminClient,
+):
+    topic_base = "".join(random.choices(ascii_letters, k=5))
+    prefix_1 = "ab"
+    prefix_2 = "fx"
+    new_topics = [prefix_1 + topic_base, prefix_2 + topic_base, prefix_1 + prefix_2 + topic_base]
+    for new_topic in new_topics:
+        topic_controller.create_topics([Topic(new_topic, replication_factor=1)])
```
Comment (swenzel) [inline tests/integration/commands/test_get.py]: You could create all at once with:
```python
topic_controller.create_topics([Topic(new_topic, replication_factor=1) for new_topic in new_topics])
```

File: tests/integration/commands/test_get.py [hunk 20]
```diff
@@ -1,11 +1,55 @@
...
+    confluent_admin_client.poll(timeout=1)
+
+    result = non_interactive_cli_runner.invoke(get_topics, ["-p", prefix_1])
+
+    assert result.exit_code == 0
+    retrieved_topics = result.output.split("\n")
```
Comment (swenzel) [inline tests/integration/commands/test_get.py]: You can use your own new feature here. Output it as json and parse it to get the exact list of topics.

File: tests/integration/commands/test_get.py [hunk 21]
```diff
@@ -1,11 +1,55 @@
...
+        if retrieved_topic:
+            assert retrieved_topic[: len(prefix_1)] == prefix_1
 
 
 @pytest.mark.integration
-def test_smoke_test_get_topics(cli_runner: CliRunner):
-    result = cli_runner.invoke(get_topics, ["--verbose"])
+def test_smoke_test_get_consumergroups(non_interactive_cli_runner: CliRunner):
+    result = non_interactive_cli_runner.invoke(get_consumergroups)
+    assert "No config provided in" not in result.output
```
Comment (swenzel) [inline tests/integration/commands/test_get.py]: See previous comment about this.

File: tests/integration/commands/test_get.py [hunk 22]
```diff
@@ -1,11 +1,55 @@
...
+    assert result.exit_code == 0
+
 
+@pytest.mark.integration
+def test_smoke_test_get_brokers(non_interactive_cli_runner: CliRunner):
+    result = non_interactive_cli_runner.invoke(get_brokers)
+    assert "No config provided in" not in result.output
```
Comment (swenzel) [inline tests/integration/commands/test_get.py]: See previous comment about this.

File: tests/integration/commands/test_ping.py [hunk 23]
```diff
@@ -7,17 +7,17 @@
 
 
 @pytest.mark.integration
-def test_smoke_test_ping(cli_runner: CliRunner):
-    result = cli_runner.invoke(ping, ["--verbose"])
-
+def test_smoke_test_ping(non_interactive_cli_runner: CliRunner):
+    result = non_interactive_cli_runner.invoke(ping, ["--verbose"])
+    assert "No config provided in" not in result.output
```
Comment (swenzel) [inline tests/integration/commands/test_ping.py]: See previous comment about this.

**Review by swenzel — CHANGES_REQUESTED** [review]

File: esque/cli/commands.py (hunk 7 above)
Comment (garrettthomaskth) [reply esque/cli/commands.py]: 👍

File: tests/integration/commands/test_get.py (hunk 19 above)
Comment (garrettthomaskth) [reply tests/integration/commands/test_get.py]: :

Comment (garrettthomaskth) [reply tests/integration/commands/test_get.py]: 👍

File: esque/cli/helpers.py [hunk 24]
```diff
@@ -3,10 +3,25 @@
...
+def isatty(stream) -> bool:
+    return _isatty(stream)
+
 
 def ensure_approval(question: str, *, no_verify: bool = False) -> bool:
     if no_verify:
         return True
+
+    if not isatty(click.get_text_stream("stdin")):
+        raise NoConfirmationPossibleException
```
Comment (swenzel) [inline esque/cli/helpers.py]: Please instantiate your objects!

File: esque/cli/output.py [hunk 25]
```diff
@@ -189,6 +195,7 @@ def green_bold(s: str) -> str:
 
 
 STYLE_MAPPING = {
+    "Topic": green_bold,
```
Comment (swenzel) [inline esque/cli/output.py]: I think you need `"topic"` here

File: tests/conftest.py [hunk 26]
```diff
@@ -1,11 +1,15 @@
...
 from pathlib import Path
 from string import ascii_letters
-from typing import Callable, Iterable, Tuple
+from typing import Callable, Iterable, Tuple, Dict
+from unittest import mock
 
 import confluent_kafka
 import pytest
+import yaml
+from _pytest.monkeypatch import MonkeyPatch
```
Comment (swenzel) [inline tests/conftest.py]: I think you don't need this anymore

File: tests/integration/commands/test_get.py [hunk 27]
```diff
@@ -1,11 +1,80 @@
...
+    topic_controller.create_topics([Topic(new_topic, replication_factor=1) for new_topic in new_topics])
+    confluent_admin_client.poll(timeout=1)
+
+    result = non_interactive_cli_runner.invoke(get_topics, ["-p", prefix_1, "-o", "json"])
+
+    assert result.exit_code == 0
+    retrieved_topics = json.loads(result.output)
+    assert len(retrieved_topics) > 1
+    for retrieved_topic in retrieved_topics:
+        if retrieved_topic:
```
Comment (swenzel) [inline tests/integration/commands/test_get.py]: Now that you're getting a proper list, there won't be empty lines and therefore no empty topics. You can get rid of that check.

File: tests/integration/commands/test_get.py [hunk 28]
```diff
@@ -1,11 +1,80 @@
...
+def test_smoke_test_get_consumergroups_with_output_format(
+    non_interactive_cli_runner: CliRunner, output_format: str, loader: Callable
+):
+    result = non_interactive_cli_runner.invoke(get_consumergroups, ["-o", output_format])
+    assert result.exit_code == 0
+    loader(result.output)
 
+
+@pytest.mark.integration
+def test_smoke_test_get_brokers(non_interactive_cli_runner: CliRunner):
```
Comment (swenzel) [inline tests/integration/commands/test_get.py:69]: No `loader` here?

File: tests/integration/commands/test_ping.py [hunk 29]
```diff
@@ -7,17 +7,16 @@
 
 
 @pytest.mark.integration
-def test_smoke_test_ping(cli_runner: CliRunner):
-    result = cli_runner.invoke(ping, ["--verbose"])
-
+def test_smoke_test_ping(non_interactive_cli_runner: CliRunner):
```
Comment (swenzel) [inline tests/integration/commands/test_ping.py:10]: No `loader` here either?

File: tests/integration/commands/test_get.py [hunk 30]
```diff
@@ -1,11 +1,81 @@
...
+    topic_controller.create_topics([Topic(new_topic, replication_factor=1) for new_topic in new_topics])
+    confluent_admin_client.poll(timeout=1)
+
+    result = non_interactive_cli_runner.invoke(get_topics, ["-p", prefix_1, "-o", "json"])
+
+    assert result.exit_code == 0
+    retrieved_topics = json.loads(result.output)
+    assert len(retrieved_topics) > 1
+    for retrieved_topic in retrieved_topics:
+        assert retrieved_topic[: len(prefix_1)] == prefix_1
```
Comment (swenzel) [inline tests/integration/commands/test_get.py]:
```suggestion
        assert retrieved_topic.startswith(prefix_1)
```

**Review by swenzel — APPROVED** [review]

**Pull request merged**